src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -2785,5 +2837,1166 @@ private void AssertValid()
+        public static (BigInteger Quotient, BigInteger Remainder) DivRem(BigInteger left, BigInteger right)
+        {
+            BigInteger quotient = DivRem(left, right, out BigInteger remainder);
+            return (quotient, remainder);
+        }
+
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.LeadingZeroCount(TSelf)" />
+        public static BigInteger LeadingZeroCount(BigInteger value)
+        {
+            value.AssertValid();
+
+            if (value._bits is null)
+            {
+                return int.LeadingZeroCount(value._sign);
+            }
+
+            // When the value is positive, we just need to get the lzcnt of the most significant bits
+            // Otherwise, we're negative and the most significant bit is always set.
+
+            return (value._sign >= 0) ? uint.LeadingZeroCount(value._bits[^1]) : 0;
+        }
+
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.PopCount(TSelf)" />
+        public static BigInteger PopCount(BigInteger value)
+        {
+            value.AssertValid();
+
+            if (value._bits is null)
+            {
+                return int.PopCount(value._sign);
+            }
+
+            ulong result = 0;
+
+            if (value._sign >= 0)
+            {
+                // When the value is positive, we simply need to do a popcount for all bits
+
+                for (int i = 0; i < value._bits.Length; i++)
+                {
+                    uint part = value._bits[i];
+                    result += uint.PopCount(part);
+                }
+            }
+            else
+            {
+                // When the value is negative, we need to popcount the two's complement representation
+                // We'll do this "inline" to avoid needing to unnecessarily allocate.
+
+                int i = 0;
+                uint part;
+
+                do
+                {
+                    // Simply process bits, adding the carry while the previous value is zero
+
+                    part = ~value._bits[i] + 1;
+                    result += uint.PopCount(part);
+
+                    i++;
+                }
+                while ((part == 0) && (i < value._bits.Length));
+
+                while (i < value._bits.Length)
+                {
+                    // Then process the remaining bits only utilizing the one's complement
+
+                    part = ~value._bits[i];
+                    result += uint.PopCount(part);
+
+                    i++;
+                }
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.RotateLeft(TSelf, int)" />
+        public static BigInteger RotateLeft(BigInteger value, int rotateAmount)
+        {
+            value.AssertValid();
+            int byteCount = (value._bits is null) ? sizeof(int) : (value._bits.Length * 4);
+
+            // Normalize the rotate amount to drop full rotations
+            rotateAmount %= (int)(byteCount * 8L);
```

Comment:
Is this guaranteed to handle overflow (negative _byteCount_) correctly? Some very limited testing says yes, but I wondered if you knew if this was a solid guarantee.

Reply:
I believe so, although its possible there is some edge case.

More concretely we have APIs that allow getting the underlying bytes into an array or span, and so long term I want to limit `_bits.Length` to be no more than `Array.MaxLength` bytes. This ensures you can always fit the underlying data into a span and `2^2_147_483_591` should be more than big enoguh for anyone.

The shifting and rotate APIs are then somewhat "oddities" in that they currently take `int` but theoretically we allow up to `Array.MaxLength * 8` bits, which is currently up to `17,179,868,728` bits (requiring up to 34 bits).

If we limit the actual value to no more than `int.MaxLength` bits so these APIs continue to make sense then we allow `BigInteger` up to `268_435_455` bytes or approximately 256 MB. This allows an astronomically large number of approx: `7.15 * 10^80_807_123`.

I'm not necessarily "hardset" on either of these limits, but I think they are practical and realistic for .NET. Elsewise we need "chunking" APIs and various APIs may throw or not function as expected in all scenarios.

Reply:
Added handling to ensure that `_bits.Length <= (Array.Length / sizeof(uint))` and to throw an `OverflowException` otherwise.

Changed the modulus logic to then be `rotateAmount = (int)(rotateAmount % (byteCount * 8L))`